src/field/vecops.cxx

```diff
@@ -34,69 +34,85 @@
  * Gradient operators
  **************************************************************************/
 
-const Vector2D Grad(const Field2D &f, CELL_LOC UNUSED(outloc)) {
+const Vector2D Grad(const Field2D &f, CELL_LOC outloc) {
   Vector2D result(f.getMesh());
 
   TRACE("Grad( Field2D )");
   
-  result.x = DDX(f);
-  result.y = DDY(f);
-  result.z = DDZ(f);
+  if (outloc == CELL_DEFAULT) {
+    outloc = f.getLocation();
+  }
+
+  result.x = DDX(f, outloc);
+  result.y = DDY(f, outloc);
+  result.z = DDZ(f, outloc);
 
   result.covariant = true;
 
   return result;
 }
 
-const Vector3D Grad(const Field3D &f, 
-                    CELL_LOC outloc_x, CELL_LOC outloc_y, CELL_LOC outloc_z) {
+const Vector3D Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
+                    CELL_LOC outloc_z) {
   Vector3D result(f.getMesh());
 
   TRACE("Grad( Field3D )");
 
-  if(outloc_x == CELL_DEFAULT)
+  if (outloc_x == CELL_DEFAULT) {
```

Comment:
I think this might be dangerous, currently assume that either all components are at the same location or all are staggered in direction of component. Here we're allowing some to be set from input field and others to set from argument.

Reply:
I realise you're not changing this but thought I should flag!

Reply:
Ah, I see! Yes, maybe we just shouldn't expose this as part of the public API

Reply:
Yup, maybe with an ASSERT for further safety?